### PR TITLE
add Runge Kutta algorithm 4th order

### PR DESCRIPTION
**Pull request by PrometheusPi**
This pull request adds Runge Kutta 4th order to libPMacc.

### CODE REVIEW

File: src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/**
+ * 2015 Richard Pausch
```
Comment (psychocoderHPC) [inline src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp]: Copyright is missing

Comment (PrometheusPi) [inline src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp]: @psychocoderHPC I added the Copyright

File: src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+#pragma once
+
+namespace PMacc
+{
+namespace math
+{
+
+
+template<typename F, typename V, typename T>
```
Comment (psychocoderHPC) [inline src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp]: please write `T_F` or `T_Functor`, `T_ValueType`, ... like described in the [coding guide lines](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Coding-Guide-Lines)

File: src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  public:
+    /** Runge Kutta solver 4th order
+     *
+     * Parameter:
+     * f functor with first argument time and second variables
+     * v variables of type V (can be vector type)
+     * t time 
+     * h time step
+     */
+    RungeKutta4 (F& f, V v, T t, T h) : f(f), v(v), t(t), h(h) {}
```
Comment (psychocoderHPC) [inline src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp]: Feel free to use more than one letter for the variable names we are no FORTRAN developers :-)

File: src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  
+    V next()
+    /** Calculate next time step based on the Runge Kutta
+     *  algorithm and return next variable
+     */
+    {
+      // calculate all 4 steps of the Runge Kutta 4th order
+      const V k_1 = f(t, v);
+      const V k_2 = f(t + 0.5*h, v+((0.5*h)*k_1));
```
Comment (psychocoderHPC) [inline src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp]: please use the same spacing in the equation `a + b * ( c + a)` (best is to surround  operators with white spaces)

File: src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+       
+      const V diff = ((h/6) * (k_1 + (2.*k_2) + (2.*k_3) + k_4));
+
+      const V out = (v+diff);
+  
+      return out;
+    }
+
+  private:
+    F& f; // differential equation (t, v)
```
Comment (psychocoderHPC) [inline src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp]: might be it is better to align the members `PMACC_ALIGN`

File: src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * Parameter:
+     * f functor with first argument time and second variables
+     * v variables of type V (can be vector type)
+     * t time 
+     * h time step
+     */
+    RungeKutta4 (F& f, V v, T t, T h) : f(f), v(v), t(t), h(h) {}
+  
+    V next()
```
Comment (psychocoderHPC) [inline src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp]: Is HDINLINE missing?

Comment (PrometheusPi) [inline src/libPMacc/include/math/RungeKutta/RungeKutta4.hpp]: I hope, the last push fixed all issues.

**Comment by PrometheusPi** [comment]
I checked the physics based on Lorentz force and the result look fine. The problem was in the new (not yet committed) pusher. I will remove the *WIP* flag.

**Comment by PrometheusPi** [comment]
@psychocoderHPC and @Heikman Is the clocation of the files in `libPMacc/include/math/` OK, or would be `libPMacc/include/algorithms/math` better.

**Pull request merged**